### PR TITLE
Fires burn bright

### PR DESCRIPTION
**Pull request by hubismal**
## About the PR 
Fires now visibly burn brightly (which is to say they're unshaded) whether they're a mob/entity on fire, or a hallway on fire.
This should also make the explosion rework (if that's still a thing) look nicer.

Specifically: 
- The fire overlay on a human is now unshaded.

- ~~The fire-recoloring logic in GasTileOverlaySystem is moved into a new function that a new overlay, FireTileOverlay, uses.~~ There is a new separate struct for fire overlay that contains only the relevant data, and logic for drawing fires. FireTileOverlay is essentially GasTileOverlay but with the unshaded shader (and it obviously only draws fire).

**Screenshots**
image
image

**Changelog**

:cl:
- tweak: Fires now burn brightly, meaning fires and people affected by fire are much more noticeable.

### CODE REVIEW

**Comment by mirrorcult** [comment]
doesnt gastileoverlay still draw the fire stuff

**Comment by hubismal** [comment]
> doesnt gastileoverlay still draw the fire stuff

it only draws gas now because the method in gastileoverlaysystem it uses used to account for fires and send the appropriate colors. now it doesn't, the code for coloring fires was moved into MoveNextFire.


(i think)

**Comment by hubismal** [comment]
To prove my point, here's what it looks like when FireTileOverlay is commented out from being added.

https://user-images.githubusercontent.com/47284081/152709467-dfe9e80b-3e90-42f6-9c00-febb72527a21.mp4

File: Content.Client/Atmos/Overlays/FireTileOverlay.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        public override OverlaySpace Space => OverlaySpace.WorldSpaceBelowFOV;
+        private readonly ShaderInstance _shader;
+
+        public FireTileOverlay()
+        {
+            IoCManager.InjectDependencies(this);
+
+            _gasTileOverlaySystem = EntitySystem.Get<GasTileOverlaySystem>();
+
+            IoCManager.InjectDependencies(this);
```
Comment (metalgearsloth) [inline Content.Client/Atmos/Overlays/FireTileOverlay.cs:28]: So nice it's injected twice.

File: Content.Client/Atmos/EntitySystems/GasTileOverlaySystem.cs [hunk 2]
```diff
@@ -217,6 +222,13 @@ public bool MoveNext(out (Texture Texture, Color Color) overlay)
                     return true;
                 }
 
+                overlay = default;
+                return false;
+            }
+            //MoveNext but used in FireTileOverlay
+            public bool MoveNextFire(out (Texture Texture, Color Color) overlay)
+            {
```
Comment (metalgearsloth) [inline Content.Client/Atmos/EntitySystems/GasTileOverlaySystem.cs:230]: You should split this to 2 separate struct enumerators.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by mirrorcult** [comment]
> To prove my point, here's what it looks like when FireTileOverlay is commented out from being added.
> 
> 

all good then

**Comment by Pangogie** [comment]
> I feel like people on fire should have a pointlight to actually glow. Makes for some scary maint chases.

They'd likely have to write some sort of edge finding code so the light is only on the edges for performance reasons.

**Comment by metalgearsloth** [comment]
> > I feel like people on fire should have a pointlight to actually glow. Makes for some scary maint chases.
> 
> They'd likely have to write some sort of edge finding code so the light is only on the edges for performance reasons.

Think they meant for people on fire and not the gas itself.

**Comment by Matz05** [comment]
As I understand the tiny, non-shadow-casting point lights are now able to be used extremely freely with minimal performance impact, and only a risk of 'not all of them drawing' in cases of overuse, so as long as you stick to those and not the shadow-casting kind, I _think_ light from fire tiles and burning entities should be OK?

**Comment by hubismal** [comment]
> As I understand the tiny, non-shadow-casting point lights are now able to be used extremely freely with minimal performance impact, and only a risk of 'not all of them drawing' in cases of overuse, so as long as you stick to those and not the shadow-casting kind, I _think_ light from fire tiles and burning entities should be OK?

If you just want a non-shadow casting glow effect from fire it would be functionally identical and probably less convoluted to make a new shader for fire that has a bloom effect on it.

**Comment by Matz05** [comment]
That's probably a better solution for fire tiles, yeah. Burning entities can probably get away with tiny point lights and the unshaded layer?

**Pull request merged**